### ISSUE TITLE
Paper-scale training step allocates ~1 GB/step (GC churn): transient backward buffers bypass arena & ArrayPool>2^20

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#478]
## Summary

A paper-scale (~BERT-Large) training step allocates **~1 GB of transient GC garbage per step** and triggers a gen2 collection **every step**, even for a batch=1 / seq=4 input where the actual compute is trivial. The cost is **allocation + GC, not math** — the custom SIMD never gets to be the bottleneck. This makes paper-scale model training pathologically slow (~2.5 s/step here) and is the root cause of downstream AiDotNet model-family training-invariant tests timing out (especially on net471, which GCs slower).

This is a general engine perf bug affecting **all** paper-scale training, on **both** TFMs — not specific to one model.

## Reproduction / measurement

Probed with a single `MatryoshkaEmbedding<float>` (TransformerEmbeddingNetwork: EmbeddingLayer[30522×768] + 12× TransformerEncoderLayer, hidden 1536 / FFN 3072), batch=1, seq=4, repeated `Train(input, target)` inside an outer `TensorArena`:

```
iter 1: 2148 ms, alloc=985.8 MB, gc=(3/2/1)   <- gen2 GC every step
iter 2: 2671 ms, alloc=985.8 MB, gc=(6/6/1)
...stable ~985.8 MB/step, loss decreasing correctly (training is correct)
```

~985 MB/step ≈ one full copy of the parameter/gradient set, re-allocated and thrown away every step.

## Allocation breakdown (instrumented `new T[]` sites + GC counters)

- **~90 MB/step**: `CpuEngine.TensorEmbeddingLookupBackward` allocates the embedding gradient via `new Tensor<T>(...)` (stack: `BackwardFunctions.TensorEmbeddingLookupBackward <- CompiledDelegateChain.Execute <- GradientTape.ComputeGradientsViaGraphCore`).
- **~895 MB/step**: bypasses **both** `TensorBase(int[])` and `VectorBase`'s array constructor. The only remaining large-array sources are:
  1. `TensorArena.TryRentTensor` / `TryAllocateCore` allocating `new T[]` directly (lines that `new T[totalSize]` without going through `Vector<T>`'s ctor), i.e. the arena allocating **fresh every step instead of reusing** across steps; and/or
  2. `TensorAllocator.RentUninitialized` → `ArrayPool<T>.Shared.Rent(n)` for **n > 2²⁰**, which `ArrayPool.Shared` does **not** retain (its largest bucket is 2²⁰ elements = 4 MB for float) — so `Rent` returns a fresh array and `Return` drops it. Every paper-scale weight gradient (embedding 30522×768, FFN 1536×3072, QKVO 1536×1536) exceeds 2²⁰, so each is GC-allocated and collected every step.

Key path: `engine.TensorMatMul` (backward gradA/gradB) → `AutoTensorCache.RentOrAllocate` → `TensorAllocator.RentUninitialized`, which tries `TensorArena.Current` first and falls to `ArrayPool.Shared.Rent` otherwise. So the transient gradient/activation buffers are only pooled if (a) the arena is active during backward AND (b) the arena actually reuses them across steps. Neither is currently happening at paper scale.

## Why existing pooling doesn't catch it

- `ArrayPool<T>.Shared` caps retained buffers at 2²⁰ elements — paper-scale gradients are far larger, so they are never pooled.
- `TensorArena` exists and has a `Reset()` "zero-alloc after warmup" design, but it pools per **exact size** in a 32-slot ring (`MaxTensorRingSlots`); a 12-layer transformer can exhaust the ring with forward-activation shapes before the large backward shapes arrive (mitigated by an uncapped dict fallback — see draft below), and the arena's pool is dropped on `Dispose`. A persistent-arena experiment (one arena reused across `Train()` calls via a new `Activate()`/`Reset()` scope instead of `Create()`/`Dispose()` per step) only recovered ~10% of the churn — confirming the bulk of the backward's large buffers are **not** flowing through the arena's reuse path.

## Recommended fix (two complementary parts)

1. **Make the backward run inside a reusable arena and confirm reuse.** Provide a persistent-arena API (`TensorArena.CreatePersistent()` + an `Activate()` scope that restores the previous current arena and `Reset()`s — rewinding cursors but **keeping** buffers) so the training loop reuses one arena across steps. Then ensure the large backward gradient/activation buffers actually hit `arena.TryRentTensor`/`TryAllocateCore` (not `ArrayPool`) — i.e. determine why they currently bypass arena reuse and fix it. When the ring is full, fall back to the arena's **uncapped dictionary pool** rather than returning null (which sends the caller to the un-pooling `ArrayPool.Shared`).
2. **Back `ArrayPool`-tier large buffers with a real pool** (safety net for non-arena callers): route `TensorAllocator` Rent/Return of `> 2²⁰`-element buffers through a size-capped large-buffer pool (or `ThreadLocalTensorCache`, which has no element cap) so `Return` actually retains them and `Rent` reuses them — instead of `ArrayPool.Shared` silently dropping them.

Target: paper-scale training step → **near-zero GC allocation after warmup** (PyTorch's caching-allocator behavior), eliminating the per-step gen2 GC.

## Diagnostic tooling used

A `tools/MemProbe` harness (times `Train()` steps, reports per-step MB + gen0/1/2) plus a temporary `AllocDiag` (counts `new T[]` by call-site/stack, tags arena vs ArrayPool paths). Can be contributed if useful for regression-guarding the fix.

## Related (separate) build break found

`CpuEngine.Mlp.cs` uses `Math.Clamp`, which does not exist on **net471** (the repo has manual clamp helpers elsewhere for exactly this reason). This breaks the net471 build of the current perf branch. Fix: `Math.Max(1, Math.Min(rows / 16, Environment.ProcessorCount))`. Worth its own small PR.

**Comment by ooples** [comment]
## Related finding from AiDotNet #1670: large-model **forward** is memory-footprint-bound, and `double` is catastrophic at scale

While profiling the AiDotNet #1670 WhisperTimestamped timeout I gathered data that's directly relevant to this GC-churn / large-footprint issue — but on the **forward/inference** path (this issue was scoped to backward buffers), and with a strong **precision** component. Flagging here per request; may warrant reopening or a new forward-path issue.

### Setup
- Model: WhisperTimestamped large-v3 scale, **631M params**, input `[1,64,32]` (M = batch×seq = 64).
- All runs `AIDOTNET_DISABLE_GPU=1` (CPU-only, to match CI). Steady-state = 2nd+ forward. Min-of-N for micro-benchmarks.

### Headline: `double` forward is ~64× slower than `float` at scale — and it's footprint, not compute

| scale | params | float fwd (steady) | double fwd (steady) | double/float |
|---|---|---|---|---|
| small (dim 256, 4+4 layers) | 3.2M | 0.040s | 0.051s | **1.3×** |
| large-v3 (dim 1280, 32+32) | 631M | 2.41s | **153s** | **64×** |

The ratio collapses from 64× to 1.3× when the working set shrinks → the gap is **memory-footprint bound** (5GB double working set vs 2.5GB float thrashing cache + driving GC), **not op compute**. Corroborating:

- **fp64 GEMM kernel is fine.** Isolated min-of-N at the hot shapes: at M=64, fp64 is *comparable to or faster than* fp32 (FFN 64×1280×5120: fp32 74ms / fp64 32ms parallel). fp32's expected 2× edge only appears at large M (512: 242 vs 122 GF/s).
- **Per-op fp64/fp32 floor ≤ 5.2×** (FusedLinearGELU 3.4×, SDPA 5.2×, LayerNorm 1.9×, GELU 1.9×) — nowhere near 64×.
- **`double`-only steady-state degradation**: forward #1 = 29s, forward #2 = 153s (5× worse). Absent in float (stays ~2.4s). Classic large-heap Gen2 GC-churn signature — the same family of problem as this issue, but triggered on the forward by the 5GB resident + transient footprint.

### Diagnostic harnesses added (branch `perf/653-small-m-no-overparallel`)
- `Benchmarks --fp64-gemm-floor` — fp32-vs-fp64 GEMM at the hot shapes.
- `Benchmarks --fp64-op-floor` — fp32-vs-fp64 for FusedLinear / LayerNorm / SDPA / GELU.
- (AiDotNet side) `testconsole whispertimestamped-profile[-f32]` with `WT_SCALE=small` for the footprint sweep.

### Takeaway
The levers for large-model forward perf are **footprint reduction** (use `float`; reduce transient/resident allocation so the working set + GC stay bounded), not op-kernel micro-optimization. This extends the backward-buffer scope of this issue to the forward/inference path and adds the precision dimension. Happy to open a dedicated forward-path issue if preferred.
